File: Server/tests/integration/game_flow_test.go
package integration

import (
	"context"
	"testing"
	"time"

	"google.golang.org/protobuf/proto"
	pb "tienlen/proto"
)

func TestFullGameStart(t *testing.T) {
	// 1. Create 4 Clients
	clients := make([]*TestClient, 4)
	for i := 0; i < 4; i++ {
		clients[i] = NewTestClient(t)
		defer clients[i].Close()
	}
	t.Log("Created 4 clients")

	// 2. Client 0 creates a match (via find_match RPC which creates if none found)
	// We use the helper which handles finding/creating.
	matchID := clients[0].FindAndJoinMatch(t)
	t.Logf("Client 0 created/joined match: %s", matchID)

	// 3. Other clients join the SAME match
	for i := 1; i < 4; i++ {
		_, err := clients[i].Socket.JoinMatch(context.Background(), nil, matchID, nil)
		if err != nil {
			t.Fatalf("Client %d failed to join match: %v", i, err)
		}
		t.Logf("Client %d joined match", i)
	}

	// Wait a bit for presences to sync
	time.Sleep(1 * time.Second)

	// 4. Client 0 (Owner) sends StartGame
	// OpCode 1 is START_GAME
	startReq := &pb.StartGameRequest{}
	bytes, _ := proto.Marshal(startReq)

	t.Log("Client 0 sending StartGame...")
	_, err := clients[0].Socket.SendMatchState(context.Background(), matchID, 1, bytes, nil)
	if err != nil {
		t.Fatalf("Failed to send StartGame: %v", err)
	}

	// 5. Assert: All clients receive GameStarted event (OpCode 101)
	// OpCode 101 = OP_CODE_GAME_STARTED (from proto)
	const OpCodeGameStarted = 101

	// We'll use a WaitGroup logic or just check sequentially since they all should get it.
	// For simplicity, let's wait for Client 0 to get it first.
	
	for i, c := range clients {
		t.Logf("Waiting for GameStarted on Client %d...", i)
		data := c.WaitForMatchState(t, OpCodeGameStarted, 5*time.Second)
		
		var event pb.GameStartedEvent
		if err := proto.Unmarshal(data.Data, &event); err != nil {
			t.Errorf("Client %d failed to unmarshal GameStarted: %v", i, err)
			continue
		}
		
		if len(event.Hand) != 13 {
			t.Errorf("Client %d expected 13 cards, got %d", i, len(event.Hand))
		}
		t.Logf("Client %d received Hand: %v", i, event.Hand)
	}
	
	t.Log("TestPassed: Game started successfully with 4 players.")
}

File: Server/tests/integration/game_flow.test.ts
import { createTestUser, TestUser } from "./helpers";
import * as protobuf from "protobufjs";
import * as path from "path";

const OP_CODE_START_GAME = 1;
const OP_CODE_GAME_STARTED = 100; // Server -> Client event

describe("Tien Len Integration Tests", () => {
    let users: TestUser[] = [];
    let FindMatchResponse: protobuf.Type;

    beforeAll(async () => {
        const root = await protobuf.load(path.join(__dirname, "../../../proto/tienlen.proto"));
        FindMatchResponse = root.lookupType("tienlen.v1.FindMatchResponse");
    });

    afterEach(async () => {
        // Cleanup all users
        await Promise.all(users.map(u => u.close()));
        users = [];
    });

    test("Full Game Start Flow (4 Players)", async () => {
        // 1. Create 4 Users
        console.log("Creating 4 users...");
        for (let i = 0; i < 4; i++) {
            users.push(await createTestUser());
        }

        const owner = users[0];
        console.log(`User 0 (${owner.userId}) initialized.`);

        // 2. Owner Finds/Creates Match via Test RPC
        console.log("Owner calling test_create_match...");
        const rpcResult = await owner.client.rpc(owner.session, "test_create_match", {});
        
        // Nakama JS automatically parses the quoted string from server into a plain string.
        const finalMatchId = rpcResult.payload as unknown as string;
        
        if (!finalMatchId) {
            throw new Error(`RPC test_create_match returned empty matchId`);
        }
        
        console.log(`Match created: ${finalMatchId}`);
        
        // Owner must join
        await owner.socket.joinMatch(finalMatchId);
        console.log("Owner joined match.");

        // 3. Other players join
        for (let i = 1; i < 4; i++) {
            await users[i].socket.joinMatch(finalMatchId);
            console.log(`User ${i} joined match.`);
        }

        // Wait a bit for presences to sync
        await new Promise(r => setTimeout(r, 1000));

        // 4. Setup Listeners for GameStarted
        const startPromises = users.map(u => u.waitForOpCode(OP_CODE_GAME_STARTED));

        // 5. Owner starts game
        console.log("Owner sending StartGame...");
        await owner.socket.sendMatchState(finalMatchId, OP_CODE_START_GAME, new Uint8Array(0));

        // 6. Assert everyone got the event
        await Promise.all(startPromises);
        console.log("All players verified Game Started.");
    }, 15000); // 15s timeout
});
